File: store/src/main/java/com/nytimes/android/external/store3/util/NoopPersister.java
package com.nytimes.android.external.store3.util;

import com.nytimes.android.external.cache3.Cache;
import com.nytimes.android.external.cache3.CacheBuilder;
import com.nytimes.android.external.store3.base.Clearable;
import com.nytimes.android.external.store3.base.Persister;
import com.nytimes.android.external.store3.base.impl.MemoryPolicy;

import java.util.concurrent.TimeUnit;
import javax.annotation.Nonnull;

import io.reactivex.Maybe;
import io.reactivex.Single;

/**
 * Pass-through diskdao for stores that don't want to use persister
 */
public class NoopPersister<Raw, Key> implements Persister<Raw, Key>, Clearable<Key> {
    protected final Cache<Key, Maybe<Raw>> networkResponses;

    NoopPersister(MemoryPolicy memoryPolicy) {
        if (memoryPolicy.hasAccessPolicy()) {
            networkResponses = CacheBuilder.newBuilder()
                .expireAfterAccess(memoryPolicy.getExpireAfterAccess(), memoryPolicy.getExpireAfterTimeUnit())
                .build();

        } else if (memoryPolicy.hasWritePolicy()) {
            networkResponses = CacheBuilder.newBuilder()
                .expireAfterWrite(memoryPolicy.getExpireAfterWrite(), memoryPolicy.getExpireAfterTimeUnit())
                .build();
        } else {
            throw new IllegalArgumentException("No expiry policy set on memory-policy.");
        }
    }

    public static <Raw, Key> NoopPersister<Raw, Key> create() {
        return NoopPersister.create(null);
    }

    public static <Raw, Key> NoopPersister<Raw, Key> create(MemoryPolicy memoryPolicy) {
        if (memoryPolicy == null) {
            MemoryPolicy defaultPolicy = MemoryPolicy
                .builder()
                .setExpireAfterWrite(24)
                .setExpireAfterTimeUnit(TimeUnit.HOURS)
                .build();
            return new NoopPersister<>(defaultPolicy);
        }
        return new NoopPersister<>(memoryPolicy);
    }

    @Nonnull
    @Override
    public Maybe<Raw> read(@Nonnull Key key) {
        Maybe<Raw> cachedValue = networkResponses.getIfPresent(key);
        return cachedValue == null ? Maybe.<Raw>empty() : cachedValue;
    }

    @Nonnull
    @Override
    public Single<Boolean> write(@Nonnull Key key, @Nonnull Raw raw) {
        networkResponses.put(key, Maybe.just(raw));
        return Single.just(true);
    }

    @Override
    public void clear(@Nonnull Key key) {
        networkResponses.invalidate(key);
    }
}

File: store/src/main/java/com/nytimes/android/external/store3/util/NoopPersister.kt
package com.nytimes.android.external.store3.util

import com.nytimes.android.external.cache3.Cache
import com.nytimes.android.external.cache3.CacheBuilder
import com.nytimes.android.external.store3.base.Clearable
import com.nytimes.android.external.store3.base.Persister
import com.nytimes.android.external.store3.base.impl.MemoryPolicy
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import java.util.concurrent.TimeUnit

/**
 * Pass-through diskdao for stores that don't want to use persister
 */
class NoopPersister<Raw, Key> internal constructor(memoryPolicy: MemoryPolicy) : Persister<Raw, Key>, Clearable<Key> {
    val networkResponses: Cache<Key, Deferred<Raw>>
    private val memoryScope = CoroutineScope(SupervisorJob())

    init {
        if (memoryPolicy.hasAccessPolicy()) {
            networkResponses = CacheBuilder.newBuilder()
                    .expireAfterAccess(memoryPolicy.expireAfterAccess, memoryPolicy.expireAfterTimeUnit)
                    .build()

        } else if (memoryPolicy.hasWritePolicy()) {
            networkResponses = CacheBuilder.newBuilder()
                    .expireAfterWrite(memoryPolicy.expireAfterWrite, memoryPolicy.expireAfterTimeUnit)
                    .build()
        } else {
            throw IllegalArgumentException("No expiry policy set on memory-policy.")
        }
    }

    override suspend fun read(key: Key): Raw? = networkResponses.getIfPresent(key)?.await()


    override suspend fun write(key: Key, raw: Raw): Boolean {
        networkResponses.put(key, memoryScope.async { raw })
        return true
    }

    override fun clear(key: Key) {
        networkResponses.invalidate(key)
    }

    companion object {

        fun <Raw, Key> create(): NoopPersister<Raw, Key> {
            return NoopPersister.create(null)
        }

        fun <Raw, Key> create(memoryPolicy: MemoryPolicy?): NoopPersister<Raw, Key> {
            if (memoryPolicy == null) {
                val defaultPolicy = MemoryPolicy
                        .builder()
                        .setExpireAfterWrite(24)
                        .setExpireAfterTimeUnit(TimeUnit.HOURS)
                        .build()
                return NoopPersister(defaultPolicy)
            }
            return NoopPersister(memoryPolicy)
        }
    }
}
